passage/src/main/java/id/passage/android/PassageHosted.kt

```diff
@@ -0,0 +1,177 @@
+package id.passage.android
+
+import android.app.Activity
+import android.net.Uri
+import androidx.browser.customtabs.CustomTabsIntent
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import id.passage.android.exceptions.HostedAuthorizationError
+import id.passage.android.model.AuthResult
+import id.passage.client.infrastructure.ClientException
+import id.passage.client.infrastructure.ServerException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.toRequestBody
+import java.net.URLEncoder
+import java.security.MessageDigest
+import java.security.SecureRandom
+import java.util.Base64
+
+internal class PassageHosted {
+    internal companion object {
+        private var verifier = ""
+        private var state = ""
+        private const val CODE_CHALLENGE_METHOD = "S256"
+        private const val SECRET_STRING_LENGTH = 32
+        private val basePathOIDC = "https://${Passage.authOrigin}"
+        private val appId = Passage.appId
+        private val packageName = Passage.packageName
+
+        internal fun openChromeTab(activity: Activity) {
+            val redirectUri = "$basePathOIDC/android/$packageName/callback"
+            state = getRandomString()
+            val randomString = getRandomString()
+            verifier = getRandomString()
+            val codeChallenge = sha256Hash(randomString)
+            val newParams =
+                listOf(
+                    "client_id" to appId,
+                    "redirect_uri" to redirectUri,
+                    "state" to state,
+                    "code_challenge" to codeChallenge,
+                    "code_challenge_method" to CODE_CHALLENGE_METHOD,
+                    "scope" to "openid",
+                    "response_type" to "code",
+                ).joinToString("&") { (key, value) ->
+                    "$key=${URLEncoder.encode(value, "UTF-8")}"
+                }
+            val url = "$basePathOIDC/authorize?$newParams"
+            val intent = CustomTabsIntent.Builder().build()
+            intent.launchUrl(activity, Uri.parse(url))
+        }
+
+        private fun getRandomString(): String {
```

Comment:
Are `getRandomString` and `sha256Hash` methods duplicates from `PassageSocial`? If so, could we figure out a way to share these methods across the classes, rather than have duplicate code?